src/libraries/System.Private.Uri/src/System/Uri.cs

```diff
@@ -4207,33 +4198,34 @@ private unsafe Check CheckCanonical(char* str, ref int idx, int end, char delim)
 
                         if (char.IsHighSurrogate(c))
                         {
-                            if ((i + 1) < end)
+                            if ((uint)(i + 1) < (uint)span.Length)
                             {
-                                valid = IriHelper.CheckIriUnicodeRange(c, str[i + 1], out _, true);
+                                valid = IriHelper.CheckIriUnicodeRange(c, span[i + 1], out _, true);
+                                i++;
```

Comment:
This was a bug in previous logic - since we didn't increment `i` when encountering a surrogate pair, the next iteration would see an unpaired low surrogate and mark the input as non-iri-canonical.